jans-cedarling/cedarling/src/jwt/mod.rs

```diff
@@ -96,31 +101,60 @@ impl JwtService {
-
         // extract claims without validation
-        let access_token_claims = self.decoding_strategy.extract_claims(access_token)?;
-        let id_token_claims = self.decoding_strategy.extract_claims(id_token)?;
-        let userinfo_token_claims = self.decoding_strategy.extract_claims(userinfo_token)?;
+        let access_token_claims = DecodingStrategy::extract_claims(access_token)?;
+        let id_token_claims = DecodingStrategy::extract_claims(id_token)?;
+        let userinfo_token_claims = DecodingStrategy::extract_claims(userinfo_token)?;
 
         // validate access_token
-        let access_token = self.decoding_strategy.decode::<AccessToken>(
-            access_token,
-            None::<String>, // TODO: validate issuer for access token
-            None::<String>,
-            false,
-        )?;
+        let access_token = self
+            .decoding_strategy
+            .decode::<AccessToken>(
+                access_token,
+                None::<String>, // TODO: validate issuer for access token
+                None::<String>,
+                None::<String>,
+                true,
+                true,
+            )
+            .map_err(|e| Error::InvalidAccessToken(e.into()))?;
 
         // validate the id_token against the access_token's `iss` and `aud`
-        self.decoding_strategy.decode::<IdToken>(
-            id_token,
-            Some(&access_token.iss),
-            Some(&access_token.aud),
-            true,
-        )?;
+        let id_token = self
+            .decoding_strategy
+            .decode::<IdToken>(
+                id_token,
+                Some(&access_token.iss),
+                Some(&access_token.aud),
+                None::<String>,
+                true,
+                true,
+            )
+            .map_err(|e| Error::InvalidIdToken(e.into()))?;
+
+        // validate the userinfo_token
+        let userinfo_token = self
+            .decoding_strategy
+            .decode::<UserInfoToken>(
+                userinfo_token,
+                None::<String>,
+                None::<String>,
+                Some(id_token.sub), // validate that the `sub` is the same as with the id_token's sub
+                false,              // this token usually does not have an nbf field
+                false,              // this token usually does not have an exp field
+            )
+            .map_err(|e| Error::InvalidUserinfoToken(e.into()))?;
+        match self.decoding_strategy {
+            DecodingStrategy::WithoutValidation => (), // do nothing
+            DecodingStrategy::WithValidation { .. } => {
+                // validate that  the userinfo_token's client_id is the same as the access_token's aud
+                if userinfo_token.client_id != access_token.aud {
+                    let validation_err = Error::Validation("the userinfo_token's `client_id` does not match with the access_token's `aud`".into());
```

Comment:
We can use add another error case, instead of put to `Error::Validation` string.

Reply:
please check how i improved the error handling here: [fb93439](https://github.com/JanssenProject/jans/pull/9939/commits/fb93439d5579296a88c20a47c9fb12ba7c9dcfbc)